tests/test_xpath.cpp doesn't build on visual studio using the c++20 standard

tests/test_xpath.cpp doesn't build on Visual Studio using the C++20 standard with the following error:

```
../subprojects/pugixml-1.12/tests/test_xpath.cpp(166): error C2039: 'random_shuffle': is not a member of 'std'
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.30.30705\include\algorithm(40): note: see declaration of 'std'
../subprojects/pugixml-1.12/tests/test_xpath.cpp(166): error C3861: 'random_shuffle': identifier not found
../subprojects/pugixml-1.12/tests/test_xpath.cpp(195): error C2039: 'random_shuffle': is not a member of 'std'
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.30.30705\include\algorithm(40): note: see declaration of 'std'
../subprojects/pugixml-1.12/tests/test_xpath.cpp(195): error C3861: 'random_shuffle': identifier not found
```

I assume this also happens using C++17, but I didn't test that.

The problem appears to be that the silly compiler doesn't define the standard __cplusplus macro correctly.  See https://docs.microsoft.com/en-us/cpp/build/reference/zc-cplusplus?view=msvc-170 for details on that.  Looks like you've got workarounds for this in pugixml.hpp that use _MSC_VER.  Maybe this also needs something like that?  There's also _MSVC_LANG on VS >= 2015 update 3 which has correct __cplusplus values without requiring /Zc:__cplusplus (which actually fixes __cplusplus values on VS >= 2017 version 15.7).  See https://docs.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=msvc-170 for details on that.

[rant]I have no idea why M$ takes such pride in making everyone that tries to make something cross platform suffer so badly by ignoring standards.[/rant]

--- a/tests/test_xpath.cpp
+++ b/tests/test_xpath.cpp
@@ -10,14 +10,6 @@
 #include <algorithm>
 #include <limits>
 
-//  std::random_shuffle is deprecated in c++14, is removed in c++17.
-#if defined(__cplusplus) && (__cplusplus >= 201402L)
-#  include <random>
-#  define PUGIXML_SHUFFLE(rng)  std::shuffle(rng.begin(), rng.end(), std::default_random_engine{std::random_device{}()})
-#else
-#  define PUGIXML_SHUFFLE(rng)  std::random_shuffle(rng.begin(), rng.end())
-#endif
-
 using namespace pugi;
 
 static void load_document_copy(xml_document& doc, const char_t* text)
@@ -26,6 +18,22 @@ static void load_document_copy(xml_document& doc, const char_t* text)
 	CHECK(source.load_string(text));
 
 	doc.append_copy(source.first_child());
+}
+
+template <typename T>
+static void random_shuffle(std::vector<T>& v)
+{
+	size_t rng = 2147483647;
+
+	for (size_t i = v.size() - 1; i > 0; --i)
+	{
+		// Fisher-Yates shuffle
+		size_t j = rng % (i + 1);
+		std::swap(v[j], v[i]);
+
+		// LCG RNG, constants from Numerical Recipes
+		rng = rng * 1664525 + 1013904223;
+	}
 }
 
 TEST(xpath_allocator_many_pages)
@@ -163,7 +171,7 @@ TEST(xpath_sort_random_medium)
 	xpath_node_set ns = doc.select_nodes(STR("//node() | //@*"));
 
 	std::vector<xpath_node> nsv(ns.begin(), ns.end());
-	PUGIXML_SHUFFLE(nsv);
+	random_shuffle(nsv);
 
 	xpath_node_set copy(&nsv[0], &nsv[0] + nsv.size());
 	copy.sort();
@@ -192,7 +200,7 @@ TEST(xpath_sort_random_large)
 	xpath_node_set ns = doc.select_nodes(STR("//node() | //@*"));
 
 	std::vector<xpath_node> nsv(ns.begin(), ns.end());
-	PUGIXML_SHUFFLE(nsv);
+	random_shuffle(nsv);
 
 	xpath_node_set copy(&nsv[0], &nsv[0] + nsv.size());
 	copy.sort();

